rust/user-info-fetcher/src/backend/active_directory.rs

```diff
@@ -0,0 +1,394 @@
+    )
+    .await
+}
+
+#[tracing::instrument(skip(ldap, base_dn, user, custom_attribute_mappings), fields(user.dn))]
+async fn user_attributes(
+    ldap: &mut Ldap,
+    base_dn: &str,
+    user: &SearchEntry,
+    custom_attribute_mappings: &BTreeMap<String, String>,
+) -> Result<UserInfo, Error> {
+    let user_sid = user
+        .bin_attrs
+        .get(LDAP_FIELD_OBJECT_SECURITY_ID)
+        .into_iter()
+        .flatten()
+        .next()
+        .map(|sid| SecurityId::from_bytes(sid).context(ParseUserSidSnafu { user_dn: &user.dn }))
+        .transpose()?;
+    let id = user
+        .bin_attrs
+        .get(LDAP_FIELD_OBJECT_ID)
+        .and_then(|values| values.first())
+        .map(|uuid|
+             // AD stores UUIDs as little-endian bytestrings
+             // Technically, byte order doesn't matter to us as long as it matches the filter, but
+             // we should try to be consistent with how MS tools display the UUIDs
+             Uuid::from_slice_le(uuid).context(ParseIdByLdapSnafu))
+        .transpose()?;
+    let username = user
+        .attrs
+        .get(LDAP_FIELD_USER_NAME)
+        .and_then(|values| values.first())
+        .cloned();
+    let custom_attributes = custom_attribute_mappings
+        .iter()
+        .filter_map(|(uif_key, ldap_key)| {
+            Some((
+                uif_key.clone(),
+                serde_json::Value::Array(match ldap_key.as_str() {
+                    // Some fields require special handling
+                    LDAP_FIELD_OBJECT_DISTINGUISHED_NAME => {
+                        vec![serde_json::Value::String(user.dn.clone())]
+                    }
+                    LDAP_FIELD_OBJECT_ID => {
+                        vec![serde_json::Value::String(id?.to_string())]
+                    }
+                    LDAP_FIELD_OBJECT_SECURITY_ID => {
+                        vec![serde_json::Value::String(user_sid.as_ref()?.to_string())]
+                    }
+
+                    // Otherwise, try to read the string value(s)
+                    _ => user
+                        .attrs
+                        .get(ldap_key)?
+                        .iter()
+                        .cloned()
+                        .map(serde_json::Value::String)
+                        .collect(),
+                }),
+            ))
+        })
+        .collect::<HashMap<_, _>>();
+    let groups = if let Some(user_sid) = &user_sid {
+        user_group_distinguished_names(ldap, base_dn, user, user_sid).await?
+    } else {
+        tracing::debug!(user.dn, "user has no SID, cannot fetch groups...");
+        Vec::new()
+    };
+
+    Ok(UserInfo {
+        id: id.map(|id| id.to_string()),
+        username,
+        groups,
+        custom_attributes,
+    })
+}
+
+/// Gets the distinguished names of all of `user`'s groups, both primary and secondary.
+#[tracing::instrument(skip(ldap, base_dn, user, user_sid))]
+async fn user_group_distinguished_names(
```

Comment:
Would it make sense to add tests for this at some point? Doesn't need to be before the release, I dragged my feet reviewing this long enough to make that non-viable any more, but maybe in a later PR?

Reply:
We need to think about nice ways to set it up, but ideally yes.

Reply:
I was kinda hoping that we'd be able to find a list of a few hundred known special and edge cases with proper parsing that we can just run this against as a unit test ..
I'll see myself out :)

Reply:
https://github.com/stackabletech/opa-operator/issues/640